src/java/com/zimbra/oauth/handlers/impl/GoogleContactsImport.java

```diff
@@ -0,0 +1,850 @@
+import java.text.DateFormat;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Calendar;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+
+import org.apache.commons.httpclient.methods.GetMethod;
+import org.apache.commons.lang.StringUtils;
+import org.apache.http.client.utils.URIBuilder;
+import org.xml.sax.SAXException;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.zimbra.common.service.ServiceException;
+import com.zimbra.common.util.HtmlBodyTextExtractor;
+import com.zimbra.common.util.ZimbraLog;
+import com.zimbra.cs.account.DataSource;
+import com.zimbra.cs.account.DataSource.DataImport;
+import com.zimbra.cs.account.Provisioning;
+import com.zimbra.cs.mailbox.Contact;
+import com.zimbra.cs.mailbox.Mailbox;
+import com.zimbra.cs.mime.ParsedContact;
+import com.zimbra.cs.service.mail.CreateContact;
+import com.zimbra.cs.service.util.ItemId;
+import com.zimbra.oauth.handlers.impl.GoogleOAuth2Handler.GoogleConstants;
+import com.zimbra.oauth.models.OAuthInfo;
+import com.zimbra.oauth.utilities.Configuration;
+import com.zimbra.oauth.utilities.OAuth2Constants;
+import com.zimbra.oauth.utilities.OAuth2Utilities;
+import com.zimbra.oauth.utilities.OAuthDataSource;
+
+/**
+ * The GoogleContactsImport class.<br>
+ * Used to sync contacts from the Google social service.<br>
+ * Based on the original YahooContactsImport class by @author Greg Solovyev.
+ *
+ * @author Zimbra API Team
+ * @package com.zimbra.oauth.handlers.impl
+ * @copyright Copyright © 2018
+ */
+public class GoogleContactsImport implements DataImport {
+
+    /**
+     * The datasource under import.
+     */
+    private final DataSource mDataSource;
+
+    /**
+     * Configuration wrapper.
+     */
+    private final Configuration config = Configuration.getDefaultConfiguration();
+
+    /**
+     * Constructor.
+     *
+     * @param datasource The datasource to set
+     */
+    public GoogleContactsImport(DataSource datasource) {
+        mDataSource = datasource;
+    }
+
+    @Override
+    public void test() throws ServiceException {
+        // list of contacts to create after parsing google responses
+        final List<ParsedContact> createList = new ArrayList<ParsedContact>();
+        // existing contacts from the datasource folder
+        final Set<String> existingContacts = new HashSet<String>();
+        // get a new access token and build the auth header
+        final String authorizationHeader = String.format("Bearer %s", refresh());
+        String respContent = "";
+        try {
+            // fetch contacts
+            final JsonNode jsonResponse = getContactsRequest(GoogleConstants.CONTACTS_URI,
+                authorizationHeader);
+            respContent = jsonResponse.toString();
+            if (jsonResponse != null && jsonResponse.isContainerNode()) {
```

Comment:
here null check is added for jsonResponse, which indicates it can be null, then NPE will occur at above line 152.

Reply:
Good catch.

Reply:
That is weird.